Imxrt1064 enablement

jhqian opened the pull request:
Enable Zephyr on i.MXRT1064

- Add soc files (header/svd/clock/iomux)

- Add board support for mimxrt1064_evk, the following samples have been tested:
hello_world, philosophers, synchronization, basic/blinky, and basic/button.

jhqian committed:
arch: arm: nxp: imxrt1064: add device support

- Add SoC information for RT1064

Signed-off-by: Ryan QIAN <jianghao.qian@nxp.com>

jhqian committed:
boards: arm: Add board support for mimxrt1064_evk

Add board support files for mimxrt1064_evk, the development board for
i.MXRT1064 (CM7) SoC.

- Add pinmux, dts, doc.
- Code can be loaded to SRAM.
- Tested samples: hello_world, philosophers, synchronization,
basic/blinky, and basic/button.

Signed-off-by: Ryan QIAN <jianghao.qian@nxp.com>

dbkinder commented on boards/arm/mimxrt1064_evk/doc/mimxrt1064_evk.rst:
```diff
@@ -0,0 +1,197 @@
+synchronous parallel NAND/NOR/PSRAM controller. The i.MX RT1064 runs on the
+Arm® Cortex-M7® core at 600 MHz.
+
+.. image:: mimxrt1064_evk.jpg
```
The image file in this PR is named ``mimxrt1064.jpg``, rename it or fix this reference.

dbkinder commented on boards/arm/mimxrt1064_evk/doc/mimxrt1064_evk.rst:
```diff
@@ -0,0 +1,197 @@
+
+The Segger J-Link firmware does not support command line flashing, therefore
+the usual ``flash`` build system target is not supported.
+Instead, see the https://www.nxp.com/docs/en/application-note/AN12108.pdf for flashing instructions.
```
Better would be:
```suggestion
Instead, see the NXP `How to Enable Boot from QSPI Flash App Note
<https://www.nxp.com/docs/en/application-note/AN12108.pdf>`_ for flashing instructions.
```

jhqian replied:
Will do. thanks.

dbkinder reviewed (changes requested):
Thanks for providing documentation.

nashif reviewed (dismissed):
Do we really need to import

ext/hal/nxp/mcux/devices/MIMXRT1064/MIMXRT1064.xml ?

Where is this file being used?

nashif commented:
> @nashif , it is SVD file. Do you want it removed?

well, this file is 10Mb and is not being used by Zephyr, so I would remove it.

jhqian commented:
> > @nashif , it is SVD file. Do you want it removed?
> 
> well, this file is 10Mb and is not being used by Zephyr, so I would remove it.

All right. @MaureenHelm, if you're okay to remove it, I'll do it, and perhaps you might want to update the script to import MCUX SDK to Zephyr as well.

nashif commented:
> Exception when running scancode

this is due to the large xml file. I have modified the checks to deal with errors in a different way now.

jhqian commented:
> > Exception when running scancode
> 
> this is due to the large xml file. I have modified the checks to deal with errors in a different way now.

Thank you, @nashif. Do you know the complaint about the license? I'm not quite sure what is violating this rule in my patches?

nashif commented:
> Thank you, @nashif. Do you know the complaint about the license? I'm not quite sure what is violating this rule in my patches?

There is no complaint, the license checker just times out parsing the huge XML file :)

jhqian commented:
> > Thank you, @nashif. Do you know the complaint about the license? I'm not quite sure what is violating this rule in my patches?
> 
> There is no complaint, the license checker just times out parsing the huge XML file :)

Got it! Thanks.

MaureenHelm commented:
> well, this file is 10Mb and is not being used by Zephyr, so I would remove it.

It is used in [Eclipse Debugging](https://docs.zephyrproject.org/latest/application/application.html#eclipse-debugging) to debug memory-mapped peripheral registers. I've imported SVDs for all NXP SoCs, and Nordic has started doing the same. Not sure offhand if ST, TI include it too.

galak commented on boards/arm/mimxrt1064_evk/board.h:
```diff
@@ -0,0 +1,12 @@
+/*
```
Remove board.h

jhqian replied:
Sure, will do.

galak reviewed (changes requested):
board.h isn't needed, please remove it.

jhqian commented:
@MaureenHelm & @nashif , do we have a conclusion on the SVD file? Please let me know if it shall be removed. 

By the way, @nashif , it seems I'm still having the license issue complaint which I still have no clew.

jhqian commented:
> @MaureenHelm & @nashif , do we have a conclusion on the SVD file? Please let me know if it shall be removed.
> 
> By the way, @nashif , it seems I'm still having the license issue complaint which I still have no clew.

any update for this PR? any other change I need to make?

MaureenHelm commented on boards/arm/mimxrt1064_evk/mimxrt1064_evk.yaml:
```diff
@@ -0,0 +1,15 @@
+#
+
+identifier: mimxrt1060_evk
+name: NXP MIMXRT1060-EVK
```
s/1060/1064/ in the identifier and name

jhqian replied:
done

MaureenHelm reviewed (changes requested):
Please add RT1064 to `ext/hal/nxp/mcux/README`

galak reviewed (commented):
Is there a reason we aren't supporting hyperflash on this board?  Seems similar to the 105x

galak commented on boards/arm/mimxrt1064_evk/Kconfig:
```diff
@@ -0,0 +1,21 @@
+config CODE_ITCM
+	bool "Link code into internal instruction tightly coupled memory (ITCM)"
+
+config CODE_INTERNAL_QSPI
```
Why is this 'INTERNAL'?  What is internal about it?

Why aren't we matching how the IMX 105/60 EVKs are done?

jhqian replied:
Since the flash it uses to boot from is internal, not like 105x and 106x which are using the external flash on the board.

galak commented on boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts:
```diff
@@ -0,0 +1,74 @@
+
+&flexspi0 {
+	qspi1: qspi@70000000 {
+		/* WINBOND */
```
Why windbond?  From schematics looks like an ISSI part?

jhqian replied:
since it is using the internal flash die from Winbond

galak reviewed (changes requested):
The QSPI/hyperflash support needs cleaning up, should be similar to what we have on the 1050/60 EVK.

galak commented on boards/arm/mimxrt1064_evk/doc/mimxrt1064_evk.rst:
```diff
@@ -0,0 +1,198 @@
+   https://www.nxp.com/support/developer-resources/run-time-software/i.mx-developer-resources/mimxrt1064-evk-i.mx-rt1064-evaluation-kit:MIMXRT1064-EVK
+
+.. _MIMXRT1064-EVK User Guide:
+   https://www.nxp.com/docs/en/user-guide/PMSMFOCRT1050UG.pdf
```
This doesn't seem to be a user guide for the EVK but for something else

jhqian replied:
thanks, fixed.

jhqian commented on boards/arm/mimxrt1064_evk/doc/mimxrt1064_evk.rst:
```diff
@@ -0,0 +1,198 @@
++---------------+-----------------+---------------------------+
+| WAKEUP        | GPIO            | SW0                       |
++---------------+-----------------+---------------------------+
+
```
thanks, fixed.

dbkinder reviewed (approved):
+1 for doc changes, thanks.

jhqian committed:
ext: hal: nxp: mcux: Add device files for RT1064

Add mcux 2.4.0 drivers and device header files for mimxrt1064.

Origins: NXP MCUxpresso SDK 2.4.0
URL: mcuxpresso.nxp.com
Maintained-by: External

Signed-off-by: Ryan QIAN <jianghao.qian@nxp.com>

jhqian commented:
fix the conflict by removing driver update, since the SDK 2.5.0 has been merged.

MaureenHelm commented on boards/arm/mimxrt1064_evk/board.cmake:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+board_runner_args(jlink "--device=MCIMXRT1064")
```
Does jlink support this device yet? I don't see it in https://www.segger.com/downloads/supported-devices.php

jhqian replied:
Segger claimed to support RT1064 in jlink v6.40, however the flash loader only supports the external flash (on board). I filed a ticket requesting the support of programming the internal flash, and they'll be working on it. Anyway, the device name in this arg is not correct, I'll update it.

MaureenHelm replied:
How do you program the application to internal flash? The device name in `board.cmake` gets passed to the JLink Commander when you run `ninja flash` or `ninja debug`, so it needs to match one of the device names listed on the segger website.

If we support ITCM only for now, then you can change this to:
```
board_runner_args(jlink "--device=Cortex-M7")
```
like we had in rt1050 prior to commit 7433f75e1a87a194208b1659430a8e124dee9cb2

jhqian replied:
I was using a flash loader I created for segger. I'll change it to cortex-m7.

MaureenHelm reviewed (changes requested):
Please update `ext/hal/nxp/mcux/README`

galak commented:
Would like to see if we can hopefully get #12423 first.

jhqian commented on boards/arm/mimxrt1064_evk/Kconfig.defconfig:
```diff
@@ -0,0 +1,38 @@
+
+# Reserve space for the IVT
+config TEXT_SECTION_OFFSET
+	default 0x2000
```
sure

jhqian commented on soc/arm/nxp_imx/rt/Kconfig.defconfig.mimxrt1064:
```diff
@@ -0,0 +1,47 @@
+# Kconfig - i.MX RT1064
```
will do

jhqian committed:
boards: mimxrt1064_evk: add configurations for XIP

To enable boot header and device configuration data if
CONFIG_INTERNAL_QSPI is defined.

Signed-off-by: Ryan QIAN <jianghao.qian@nxp.com>

jhqian commented:
Add XIP support for RT1064

galak commented on boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts:
```diff
@@ -0,0 +1,75 @@
+	chosen {
+#if defined(CONFIG_CODE_ITCM)
+		zephyr,flash = &itcm0;
+#elif defined(CONFIG_CODE_INTERNAL_QSPI)
```
This needs updating based on PR #12423

jhqian replied:
updated

galak commented on boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts:
```diff
@@ -0,0 +1,75 @@
+
+&flexspi0 {
+	qspi1: qspi@70000000 {
+		/* WINBOND */
```
This needs updating based on PR #12423

jhqian replied:
updated

galak reviewed (dismissed):
Some flash clean ups needed related to PR #12423 

Also, since the windbond flash is "internal" that should move into the SoC dtsi.  We need to create a RT1064 specific .dtsi file

jhqian committed:
dts: arm: nxp_rt: add flexspi1

- Add info of flexspi1

Signed-off-by: Ryan QIAN <jianghao.qian@nxp.com>

MaureenHelm commented on boards/arm/mimxrt1064_evk/mimxrt1064_evk.yaml:
```diff
@@ -0,0 +1,15 @@
+arch: arm
+toolchain:
+  - zephyr
+  - gnuarmemb
```
Need to add `- xtools` per #12452

jhqian replied:
done

MaureenHelm commented on ext/hal/nxp/mcux/boards/evkmimxrt1064/evkmimxrt1064_flexspi_nor_config.c:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2018 NXP
+ * All rights reserved.
```
Import all the SDK one files in the same commit. The XIP/DCD commit will then just be CMake changes.

jhqian replied:
done

MaureenHelm replied:
Please also update `ext/hal/nxp/mcux/README`

MaureenHelm commented on boards/arm/mimxrt1064_evk/Kconfig.defconfig:
```diff
@@ -10,6 +10,16 @@ if BOARD_MIMXRT1064_EVK
+	def_bool y
+
+config DEVICE_CONFIGURATION_DATA
+	def_bool n
```
Why is this `def_bool n`?

jhqian replied:
If it is set to y, although the dcd will be built into the header, but if there's a debug script to initialize sdram, it will fail to continue debug after programming.

MaureenHelm replied:
It's implicitly false already, so we don't need to set it false again here.

jhqian committed:
ext: hal: rt1064: add support for XIP and DCD

- Add cmake files to enable XIP

Signed-off-by: Ryan QIAN <jianghao.qian@nxp.com>

MaureenHelm reviewed (changes requested):
@nashif @galak Please revisit

jhqian commented:
Thank you, @MaureenHelm , they've been fixed.
